File: sicstus/pandigital_numbers.pl
/*

  Pandigital numbers  in SICStus Prolog.

  From
  Albert H. Beiler "Recreations in the Theory of Numbers", quoted from
  http://www.worldofnumbers.com/ninedig1.htm
  """
  [ Chapter VIII : Digits - and the magic of 9 ]
  [ I found the same exposé in Shakuntala Devi's book
    "Figuring : The Joy of Numbers" ]

  The following curious table shows how to arrange the 9 digits so that
  the product of 2 groups is equal to a number represented by the
  remaining digits."

    12 x 483 = 5796
    42 x 138 = 5796
    18 x 297 = 5346
    27 x 198 = 5346
    39 x 186 = 7254
    48 x 159 = 7632
    28 x 157 = 4396
    4 x 1738 = 6952
    4 x 1963 = 7852
  """

  See also

  * MathWorld http://mathworld.wolfram.com/PandigitalNumber.html
  """
  A number is said to be pandigital if it contains each of the digits
  from 0 to 9 (and whose leading digit must be nonzero). However,
  "zeroless" pandigital quantities contain the digits 1 through 9.
  Sometimes exclusivity is also required so that each digit is
  restricted to appear exactly once.
  """

  * Wikipedia http://en.wikipedia.org/wiki/Pandigital_number

  * Also see my other pandigital number models
    - MiniZinc: http://www.hakank.org/minizinc/pandigital_numbers.mzn
    - Gecode/R: http://www.hakank.org/gecode_r/pandigital_numbers.rb
    - Comet   : http://www.hakank.org/comet/pandigital_numbers.co
    - ECLiPSe : http://www.hakank.org/eclipse/pandigital_numbers.ecl

  Model created by Hakan Kjellerstrand, hakank@bonetmail.com
  See also my SICStus Prolog page: http://www.hakank.org/sicstus/

*/

:-use_module(library(clpfd)).
:-use_module(library(lists)).


%
% converts a number Num to/from a list of integer List given a base Base
%
toNum2(List, Base, Num) :-
        length(List, Len),      
        length(Xs, Len),
        exp_list2(Len, Base, Xs),
        scalar_product(Xs,List,#=, Num).

%
% Exponents for toNum2: [Base^(N-1), Base^(N-2), .., Base^0],
%    e.g. exp_list2(3, 10, ExpList) -> ExpList = [100,10,1]
%
exp_list2(N, Base, ExpList) :-
        length(ExpList, N),
        ( for(I, 0, N-1), 
          fromto([], In, Out, ExpList), 
          param(Base)  do 
            B is integer(Base**I),
            Out = [B|In]
        ).


pandigital([X1,X2,X3]) :-
        
        % length of numbers
        Len1 #>= 1,
        Len1 #=< 4,
        Len2 #>= 1,
        Len2 #=< 4,
        Len3 #= 4,

        Len1 #=< Len2, % symmetry breaking
        Len1 + Len2 + Len3 #= 9,

        indomain(Len1), % Thanks to Mats Carlsson for suggesting this.

        % set length of lists
        length(X1, Len1),
        length(X2, Len2),
        length(X3, Len3), % the result

        append([X1,X2,X3], Vars),
        domain(Vars,1,9),

        % convert to number
        Base = 10,

        toNum2(X1, Base, Num1),
        toNum2(X2, Base, Num2),
        toNum2(X3, Base, Res),

        % calculate result
        Num1 * Num2 #= Res,

        all_distinct(Vars),

        % search
        labeling([ff,bisect,down], Vars),

        format('~d * ~d = ~d: ', [Num1,Num2, Res]),
        write([X1,X2,X3]),nl.
        % fd_statistics.

go :-
        findall([X1,X2,X3], pandigital([X1,X2,X3]), L),
        % pandigital(_),
        length(L, Len),
        write(L),nl,
        write(len:Len),nl.

File: cpmpy/pandigital_numbers.py
"""
Pandigital numbers in cpmpy.

From Albert H. Beiler "Recreations in the Theory of Numbers", quoted from
http://www.worldofnumbers.com/ninedig1.htm
'''
Chapter VIII : Digits - and the magic of 9 

The following curious table shows how to arrange the 9 digits so that 
the product of 2 groups is equal to a number represented by the 
remaining digits.

  12 x 483 = 5796 
  42 x 138 = 5796 
  18 x 297 = 5346 
  27 x 198 = 5346 
  39 x 186 = 7254 
  48 x 159 = 7632 
  28 x 157 = 4396 
  4 x 1738 = 6952 
  4 x 1963 = 7852
'''

See also MathWorld http://mathworld.wolfram.com/PandigitalNumber.html
A number is said to be pandigital if it contains each of the digits 
from 0 to 9 (and whose leading digit must be nonzero). However, 
'zeroless' pandigital quantities contain the digits 1 through 9. 
Sometimes exclusivity is also required so that each digit is 
restricted to appear exactly once.
''

Note: The pandigital numbers problem is thus zeroless pandigital
number and each is restricted to appear exactly once.

* Wikipedia http://en.wikipedia.org/wiki/Pandigital_number


Model created by Hakan Kjellerstrand, hakank@hakank.com
See also my cpmpy page: http://www.hakank.org/cpmpy/

"""
import sys, math
import numpy as np
from cpmpy import *
from cpmpy.solvers import *
from cpmpy_hakank import *

def pandigital_numbers(base=10, start=1, len1=1, len2=4):

    max_d = base-1
    x_len = max_d+1-start
    max_num = base**4

    # Decision variables
    num1 = intvar(0, max_num,name="num1")
    num2 = intvar(0, max_num,name="num2")
    res  = intvar(0, max_num,name="res")
    x = intvar(start, max_d,shape=x_len,name="x")

    model = Model(
                  AllDifferent(x)
                  )
    
    model += [to_num([x[i] for i in range(len1)], num1, base)]
    model += [to_num([x[i] for i in range(len1,len1+len2)], num2, base)]
    model += [to_num([x[i] for i in range(len1+len2,x_len)], res, base)]
    
    model += [num1*num2 == res]

    # no number must start with 0
    model += [x[0] > 0]
    model += [x[len1] > 0]
    model += [x[len1+len2] > 0]

    # symmetry breaking
    model += [num1 <= num2]

    solutions = []
    ss = CPM_ortools(model)
    # ss.ort_solver.parameters.search_branching = ort.PORTFOLIO_SEARCH
    # ss.ort_solver.parameters.cp_model_presolve = False
    # ss.ort_solver.parameters.linearization_level = 0
    # ss.ort_solver.parameters.cp_model_probing_level = 0
    while ss.solve():
        solutions.append(x.value())
        ss += any(x != x.value())
               
    return solutions


def print_solution(x,len1,len2,x_len):
    print("".join([str(x[i]) for i in range(len1)]), "*",end=" ")
    print("".join([str(x[i]) for i in range(len1,len1+len2)]), "=",end=" ")
    print("".join([str(x[i]) for i in range(len1+len2,x_len)]))
    

num_solutions = 0
base = 10
start = 1
x_len = base-1 + 1-start
for len1 in range(1+math.floor(x_len / 3)+1):
  for len2 in range(1+math.floor(x_len / 3)+1):
      if x_len > len1 + len2:
          sol = pandigital_numbers(base, start, len1, len2)
          for s in sol:
              num_solutions += 1
              print_solution(s, len1, len2, x_len)             

print("num_solutions:", num_solutions)
